File: Julia/graph_algorithms/03-directed-graph-acyclicity.jl
struct DirectedGraph
    nrnodes::Int
    adjacencylist::Vector{Set{Int}}

    function DirectedGraph(n::Int, edges::Vector{Tuple{Int, Int}})
        adjacencylist = map(_ -> Set{Int}(), 1:n)
        for (n₁, n₂) in edges
            push!(adjacencylist[n₁], n₂)
        end
        new(n, adjacencylist)
    end
end


mutable struct DFSState
    previsit_numbers::Vector{Int}
    postvisit_numbers::Vector{Int}
    previsit_id::Int
    postvisit_id::Int

    DFSState(n::Int) = new(zeros(Int, n), zeros(Int, n), 1, 1)
end


function depthfirstsearch(graph::DirectedGraph)::DFSState
    state = DFSState(graph.nrnodes)
    for startnode in 1:graph.nrnodes
        if !isvisited(startnode, state)
            _ = explore(startnode, state, graph)
        end
    end
    state
end


isvisited(node::Int, state::DFSState)::Bool = state.previsit_numbers[node] != 0


function explore(node::Int, state::DFSState, graph::DirectedGraph)::Set{Int}
    state.previsit_numbers[node] = state.previsit_id
    state.previsit_id += 1
    stack = Vector{Int}([node])
    component = Set{Int}(node)
    while !isempty(stack)
        lastnode = pop!(stack)
        unvisited = findunvisitedneighbour(lastnode, graph, state)
        if unvisited === nothing
            state.postvisit_numbers[lastnode] = state.postvisit_id
            state.postvisit_id += 1
        else
            state.previsit_numbers[unvisited] = state.previsit_id
            state.previsit_id += 1
            push!(stack, lastnode)
            push!(stack, unvisited)
            push!(component, unvisited)
        end
    end
    component
end


function findunvisitedneighbour(node::Int, graph::DirectedGraph, state::DFSState)::Union{Nothing, Int}
    neighbours = graph.adjacencylist[node]
    for neighbour in neighbours
        if !isvisited(neighbour, state)
            return neighbour
        end
    end
    return nothing
end


converttointlist(line) = map(x -> parse(Int, x), split(line))


function readedges(nredges)
    edges = Tuple{Int, Int}[]
    for _ in 1:nredges
        from, to = readline() |> converttointlist
        push!(edges, (from, to))
    end
    edges
end


function hascycle(graph)
    finalstate = depthfirstsearch(graph)
    for (node, neighbours) in enumerate(graph.adjacencylist)
        nodenumber = finalstate.postvisit_numbers[node]
        if any([finalstate.postvisit_numbers[n] >= nodenumber for n in neighbours])
            return true
        end
    end
    false
end


function main()
    nrnodes, nredges = readline() |> converttointlist
    edges = readedges(nredges)
    graph = DirectedGraph(nrnodes, edges)
    result = hascycle(graph)
    println(Int(result))
end


main()

File: Rust/graph_algorithms/src/bin/03_directed_graph_acyclicity.rs
use std::collections::HashMap;
use std::io;

type Component = Vec<i32>;

struct Edge {
    start: i32,
    end: i32,
}

struct DFSState {
    previsit_numbers: Vec<i32>,
    postvisit_numbers: Vec<i32>,
    previsit_id: i32,
    postvisit_id: i32,
}

impl DFSState {
    fn new(nr_nodes: i32) -> Self {
        DFSState {
            previsit_numbers: vec![0; nr_nodes as usize],
            postvisit_numbers: vec![0; nr_nodes as usize],
            previsit_id: 1,
            postvisit_id: 1,
        }
    }

    fn is_visited(&self, node: i32) -> bool {
        self.previsit_numbers[node as usize - 1] > 0
    }

    fn find_an_unvisited_neighbour(&self, node_neighbors: Option<&Vec<i32>>) -> Option<i32> {
        match node_neighbors {
            None => None,
            Some(neighbors) => neighbors.iter().find(|neighbor| !self.is_visited(**neighbor)).cloned()
        }
    }
}

struct DirectedGraph {
    nr_nodes: i32,
    adjacency_list: HashMap<i32, Vec<i32>>
}

impl DirectedGraph {
    fn from_edge_list(nr_nodes: i32, edges: Vec<Edge>) -> Self {
        let mut adjacency_list = HashMap::new();
        for Edge { start, end } in edges {
            adjacency_list.entry(start).or_insert_with(Vec::new).push(end);
        }

        DirectedGraph { nr_nodes, adjacency_list }
    }

    fn explore(&self, state: &mut DFSState, start_node: i32) -> Vec<i32> {
        state.previsit_numbers[start_node as usize - 1] = state.previsit_id;
        state.previsit_id += 1;
        let mut previsit_stack = vec![start_node];
        let mut current_component = vec![start_node];
        loop {
            let Some(last_node) = previsit_stack.pop() else {
                return current_component
            };
            let neighbors = self.adjacency_list.get(&last_node);
            match state.find_an_unvisited_neighbour(neighbors) {
                None => {
                    state.postvisit_numbers[last_node as usize - 1] = state.postvisit_id;
                    state.postvisit_id += 1;
                },
                Some(neighbor) => {
                    state.previsit_numbers[neighbor as usize - 1] = state.previsit_id;
                    state.previsit_id += 1;
                    previsit_stack.push(last_node);
                    previsit_stack.push(neighbor);
                    current_component.push(neighbor);
                }
            };
        }
    }

    fn run_dfs(&self) -> (Vec<Component>, DFSState) {
        let mut state = DFSState::new(self.nr_nodes);
        let mut components = Vec::new();
        for starting_node in 1..=self.nr_nodes {
            if !state.is_visited(starting_node) {
                let current_component = self.explore(&mut state, starting_node);
                components.push(current_component);
            }
        }

        (components, state)
    }
}

fn convert_to_int_vector(line: &str) -> Vec<i32> {
    line
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect()
}

fn read_pair() -> (i32, i32) {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();

    let xs: Vec<i32> = convert_to_int_vector(&line);
    let [a, b] = xs.as_slice() else {
        panic!("Malformed input: {}", line);
    };
    (*a, *b)
}

fn read_edges(nr_edges: usize) -> Vec<Edge> {
    let mut edges = Vec::with_capacity(nr_edges);
    for _ in 0..nr_edges {
        let (start, end) = read_pair();
        edges.push(Edge { start, end });
    }

    edges
}

fn has_cycle(graph: DirectedGraph) -> bool {
    let (_, dfs_state) = graph.run_dfs();
    for (node, neighbors) in graph.adjacency_list.iter() {
        let node_number = dfs_state.postvisit_numbers[*node as usize - 1];
        if neighbors.iter().any(|neighbor| dfs_state.postvisit_numbers[*neighbor as usize - 1] >= node_number) {
            return true
        }
    }
    false
}

fn main() {
    let (nr_nodes, nr_edges) = read_pair();
    let edge_list = read_edges(nr_edges as usize);
    let graph = DirectedGraph::from_edge_list(nr_nodes, edge_list);
    let result = has_cycle(graph);
    println!("{}", result as i32);
}
